File: src/Client/Shared/TalkCompose.razor.js
export function getGifAutocomplete(apiKey, searchTerm) {
    const lang = (navigator.languages || ["en"])[0];
    const limit = 5;
    const response = await fetch(`https://tenor.googleapis.com/v2/autocomplete?key=${apiKey}&q=${searchTerm}&locale=${lang}&limit=${limit}`);
    if (response.status != 200) {
        return [];
    } else {
        const result = await response.json();
        return result["results"];
    }
}

export function searchGif(dotNetObjectReference, apiKey, searchTerm, pos) {
    const lang = (navigator.languages || ["en"])[0];
    let url = `https://tenor.googleapis.com/v2/search?q=${searchTerm}&key=${apiKey}&media_filter=nanogif,nanowebp_transparent,nanogif_transparent&locale=${lang}`;
    if (pos) {
        url += `&pos=${pos}`;
    }
    httpGetAsync(
        url,
        tenorCallback_search,
        dotNetObjectReference,
        (!!pos) ? 'AppendGifSearch' : 'PopulateGifSearch');
    if (!pos) {
        const limit = 4;
        httpGetAsync(
            `https://tenor.googleapis.com/v2/search_suggestions?key=${apiKey}&q=${searchTerm}&locale=${lang}&limit=${limit}`,
            tenorCallback_searchSuggestion,
            dotNetObjectReference,
            'GetGifSuggestions');
    }
}

export function shareGif(apiKey, id) {
    httpGetAsync(
        `https://tenor.googleapis.com/v2/posts?key=${apiKey}&ids=${id}&media_filter=tinygif,tinywebp_transparent,tinygif_transparent&limit=1`,
        tenorCallback_post,
        dotNetObjectReference,
        'PostGif');
    const lang = (navigator.languages || ["en"])[0];
    httpGetAsync(`https://tenor.googleapis.com/v2/registershare?id=${id}&key=${apiKey}&q=${searchTerm}&locale=${lang}`);
}

export function showGifSearch(dotNetObjectReference, apiKey) {
    const lang = (navigator.languages || ["en"])[0];
    httpGetAsync(
        `https://tenor.googleapis.com/v2/categories?key=${apiKey}&locale=${lang}`,
        tenorCallback_cateogries,
        dotNetObjectReference,
        'GetGifCategories');
    const suggestion_limit = 4;
    httpGetAsync(
        `https://tenor.googleapis.com/v2/trending_terms?key=${apiKey}&locale=${lang}&limit=${suggestion_limit}`,
        tenorCallback_trending,
        dotNetObjectReference,
        'GetGifSuggestions');
}

function httpGetAsync(theUrl, callback, dotNetObjectReference, identifier) {
    fetch(theUrl)
        .then(response => {
            if (callback && response.status == 200) {
                response.json()
                    .then(result => {
                        callback(dotNetObjectReference, identifier, result);
                    });
            }
        });
}

function tenorCallback_cateogries(dotNetObjectReference, identifier, response) {
    categories = response["tags"];
    const data = categories.map((v) => {
        return {
            image: v["image"],
            name: v["name"],
            searchTerm: v["searchterm"],
        };
    });
    dotNetObjectReference.invokeMethodAsync('Tavenem.Wiki.Blazor.Client', identifier, data);
}

function tenorCallback_post(dotNetObjectReference, identifier, response) {
    gifs = response["results"];
    let title;
    let url;
    for (const gif of gifs) {
        title = gif["title"];
        const mediaFormats = gif["media_formats"];
        for (const format in mediaFormats) {
            url = mediaFormats[format]["url"];
            if (url && url.length) {
                break;
            }
        }
    }
    dotNetObjectReference.invokeMethodAsync('Tavenem.Wiki.Blazor.Client', identifier, { title, url });
}

function tenorCallback_search(dotNetObjectReference, identifier, response) {
    gifs = response["results"];
    const data = {
        next: response["next"],
        gifs: [],
    };
    for (const gif of gifs) {
        const id = gif["id"];
        let url;
        const mediaFormats = gif["media_formats"];
        for (const format in mediaFormats) {
            url = mediaFormats[format]["url"];
            if (url && url.length) {
                break;
            }
        }
        data.gifs.push({ id, url });
    }
    dotNetObjectReference.invokeMethodAsync('Tavenem.Wiki.Blazor.Client', identifier, data);
}

function tenorCallback_searchSuggestion(dotNetObjectReference, identifier, response) {
    predictions = response["results"];
    dotNetObjectReference.invokeMethodAsync('Tavenem.Wiki.Blazor.Client', identifier, predictions);
}

function tenorCallback_trending(dotNetObjectReference, identifier, response) {
    terms = response["results"];
    dotNetObjectReference.invokeMethodAsync('Tavenem.Wiki.Blazor.Client', identifier, terms);
}

File: src/Client/Shared/TalkCompose.razor.cs
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;
using System.Diagnostics.CodeAnalysis;
using Tavenem.Blazor.Framework;
using Tavenem.Wiki.Blazor.Client.Internal.Models;
using Tavenem.Wiki.Blazor.SignalR;

namespace Tavenem.Wiki.Blazor.Client.Shared;

/// <summary>
/// Composition controls for a talk page.
/// </summary>
public partial class TalkCompose : IAsyncDisposable
{
    private bool _disposedValue;
    private DotNetObjectReference<TalkCompose>? _dotNetObjectReference;
    private IJSObjectReference? _emojiModule;
    private IJSObjectReference? _module;

    /// <summary>
    /// The message ID, if this is for a reply.
    /// </summary>
    [Parameter] public string? MessageId { get; set; }

    /// <summary>
    /// Raised when a message is entered.
    /// </summary>
    [Parameter] public EventCallback<ReplyRequest> Post { get; set; }

    /// <summary>
    /// The topic ID.
    /// </summary>
    [Parameter] public string? TopicId { get; set; }

    [MemberNotNullWhen(true, nameof(TopicId), nameof(NewMessage))]
    private bool CanPost => !string.IsNullOrEmpty(TopicId)
        && !string.IsNullOrWhiteSpace(NewMessage);

    private string EmojiButtonId { get; set; } = Guid.NewGuid().ToHtmlId();

    private List<GifCategory> GifCategories { get; set; } = new();

    private List<string>? GifSuggestions { get; set; }

    private List<GifInfo> Gifs { get; set; } = new();

    [Inject] private JSRuntime JSRuntime { get; set; } = default!;

    private string? NewMessage { get; set; }

    private string? NextGifs { get; set; }

    private bool ShowFullEditor { get; set; }

    private string? SearchGifText { get; set; }

    private bool ShowGifSearch { get; set; }

    [Inject] private WikiBlazorClientOptions WikiBlazorClientOptions { get; set; } = default!;

    /// <inheritdoc/>
    protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        if (firstRender)
        {
            _dotNetObjectReference = DotNetObjectReference.Create(this);
            _module = await JSRuntime.InvokeAsync<IJSObjectReference>(
                "import",
                "./_content/Tavenem.Wiki.Blazor.Client/Shared/TalkCompose.razor.js");
            _emojiModule = await JSRuntime.InvokeAsync<IJSObjectReference>(
                "import",
                "./_content/Tavenem.Wiki.Blazor.Client/tavenem-emoji.js");

            await _module.InvokeVoidAsync(
                "showGifSearch",
                _dotNetObjectReference,
                WikiBlazorClientOptions.TenorAPIKey);
            await _emojiModule.InvokeVoidAsync(
                "initialize",
                _dotNetObjectReference,
                EmojiButtonId);
        }
    }

    /// <inheritdoc/>
    public async ValueTask DisposeAsync()
    {
        // Do not change this code. Put cleanup code in 'DisposeAsync(bool disposing)' method
        await DisposeAsync(disposing: true);
        GC.SuppressFinalize(this);
    }

    /// <summary>
    /// Performs application-defined tasks associated with freeing, releasing, or resetting
    /// unmanaged resources.
    /// </summary>
    protected virtual async ValueTask DisposeAsync(bool disposing)
    {
        if (!_disposedValue)
        {
            if (disposing)
            {
                _dotNetObjectReference?.Dispose();
                if (_emojiModule is not null)
                {
                    await _emojiModule.DisposeAsync();
                }
                if (_module is not null)
                {
                    await _module.DisposeAsync();
                }
            }

            _disposedValue = true;
        }
    }

    /// <summary>
    /// Invoked by javascript interop.
    /// </summary>
    [JSInvokable]
    public void AppendGifSearch(GifSearchResults results)
    {
        NextGifs = results.Next;
        if (results.Gifs is not null)
        {
            Gifs.AddRange(results.Gifs);
        }
    }

    /// <summary>
    /// Invoked by javascript interop.
    /// </summary>
    [JSInvokable]
    public void GetGifSuggestions(List<string> suggestions)
        => GifSuggestions = suggestions.Count > 0 ? suggestions : null;

    /// <summary>
    /// Invoked by javascript interop.
    /// </summary>
    [JSInvokable]
    public void GetGifCategories(List<GifCategory> categories) => GifCategories = categories
        .Where(x => !string.IsNullOrEmpty(x.SearchTerm)
            && !string.IsNullOrEmpty(x.Image))
        .ToList();

    /// <summary>
    /// Invoked by javascript interop.
    /// </summary>
    [JSInvokable]
    public void PopulateGifSearch(GifSearchResults results)
    {
        NextGifs = results.Next;
        Gifs = results.Gifs?.ToList() ?? new();
    }

    /// <summary>
    /// Invoked by javascript interop.
    /// </summary>
    [JSInvokable]
    public void PostEmoji(string? emoji)
    {
        if (string.IsNullOrEmpty(emoji))
        {
            return;
        }

        NewMessage ??= string.Empty;
        NewMessage += emoji;
    }

    /// <summary>
    /// Invoked by javascript interop.
    /// </summary>
    [JSInvokable]
    public void PostGif(GifInfo gif)
    {
        if (string.IsNullOrEmpty(gif?.Url))
        {
            return;
        }

        NewMessage ??= string.Empty;
        NewMessage += $"[![{gif.Title}]({gif.Url})]({gif.Url})";
    }

    private async Task<IEnumerable<KeyValuePair<string, object>>> GetGifSearchSuggestionsAsync(string input)
    {
        if (_module is null)
        {
            return Enumerable.Empty<KeyValuePair<string, object>>();
        }

        var results = await _module.InvokeAsync<IEnumerable<string>>(
            "getGifAutocomplete",
            WikiBlazorClientOptions.TenorAPIKey,
            input);
        return results.Select(x => new KeyValuePair<string, object>(x, x));
    }

    private async Task OnGetMoreGifsAsync()
    {
        if (_module is null
            || string.IsNullOrWhiteSpace(SearchGifText))
        {
            return;
        }

        var query = SearchGifText.Trim();
        await _module.InvokeVoidAsync(
            "searchGif",
            _dotNetObjectReference,
            WikiBlazorClientOptions.TenorAPIKey,
            query,
            NextGifs);
    }

    private async Task OnGifSearchAsync(string query)
    {
        if (_module is null
            || string.IsNullOrWhiteSpace(query))
        {
            return;
        }

        await _module.InvokeVoidAsync(
            "searchGif",
            _dotNetObjectReference,
            WikiBlazorClientOptions.TenorAPIKey,
            query.Trim());
    }

    private async Task OnPostAsync()
    {
        if (CanPost)
        {
            await Post.InvokeAsync(new ReplyRequest(TopicId, NewMessage, MessageId));
            NewMessage = null;
        }
    }

    private async Task OnReactAsync(string emoji)
    {
        if (CanPost)
        {
            await Post.InvokeAsync(new ReplyRequest(TopicId, emoji, MessageId));
        }
    }

    private async Task OnSetSearchGifTextAsync()
    {
        if (_module is null
            || string.IsNullOrWhiteSpace(SearchGifText))
        {
            return;
        }

        var query = SearchGifText.Trim();
        await _module.InvokeVoidAsync(
            "searchGif",
            _dotNetObjectReference,
            WikiBlazorClientOptions.TenorAPIKey,
            query);
    }

    private async Task OnPostGifAsync(GifInfo gif)
    {
        ShowGifSearch = false;

        if (_module is null)
        {
            return;
        }

        await _module.InvokeVoidAsync(
            "shareGif",
            WikiBlazorClientOptions.TenorAPIKey,
            gif.Id);
    }
}